### PR TITLE
EB-179: Configure production and development Docker image builds

### PR DESCRIPTION
**Pull request by RMCrean**
Hey :) 

We have 5 workflows, 2 of which build images and the other 3 are more testing related. 

For the 3 testing related I have set it up so that both a PR or push to main or prod triggers them to run. 

**For the 2 that build images I have now set it up so that:** 
- On PR no images are built.
- On push to prod, both images are built with tag "prod".
- On push to main both images are built with tag "dev". 
- A workflow_dispatch (manual job) can be triggered to build either image on any branch (but not a fork). The default tag for the workflow dispatch will be "dev", but we have the choice to change that if we want when triggering the workflow. 

I think this matches up with our plan. I've been testing out these settings using a personal repository and there it seems to be working nicely at least. 

Related to this is of course the configuration of the branch settings, you can see those I have setup here: https://github.com/ScilifelabDataCentre/swedgene/settings/branches 

Cheers!

### CODE REVIEW

**Comment by RMCrean** [comment]
P.S. I was looking into the creating new releases/tagging question. I think we could explore this [GH action to make draft releases](https://github.com/marketplace/actions/release-drafter) after each merge to prod. It looks very complete and is actively maintained.

File: .github/workflows/data-builder-container.yml
Comment (kwentine) [inline .github/workflows/data-builder-container.yml:5]: I would explicit those comments, because even as an insider it bends my neurons. 
Suggestions:
-  Branch to be deployed in production, through buillds tagged `prod`
-  Branch to be deployed in development, through builds tagged `dev`

File: .github/workflows/data-builder-container.yml
Comment (kwentine) [inline .github/workflows/data-builder-container.yml:3]: Since we opened the possibility to dispatch builds manually, I wonder if we could limit automated `dev` builds to when feature branches are merged into `main`. My only concern, that is not solidly grounded, is that we might trigger many builds by pushing small fixes to `main` that do not warrant re-builds (e.g. `Update README`)

Comment (kwentine) [reply .github/workflows/data-builder-container.yml:3]: And on the related note of avoiding unnecessary builds I make this note for later inquiry: if we make a documentation PR to main, is there a way to `mark` the particular PR so that merging it do not trigger a new image build ?

The most general version of this question would be: could we skip re-builds for PR that touches only a set of files (for example always avoid rebuilds of hugo site when only the `scripts` directory receives changes.

This is all food for thought, should we happen to have a light sprint or something.

File: .github/workflows/data-builder-container.yml
Comment (kwentine) [inline .github/workflows/data-builder-container.yml:65]: Neat! Is there a reason why the complete image name is now needed ?

File: .github/workflows/hugo-container.yml
Comment (kwentine) [inline .github/workflows/hugo-container.yml:1]: Seeing the similarity between the hugo container and data builder workflow files makes me realize that these images are built under the same conditions. Do you see a way of avoiding this duplication ?

**Review by kwentine — APPROVED** [review]
Thanks, as always very instructive to review, and conjuring up ideas that are not blocking but maybe worth pondering on :)

File: .github/workflows/data-builder-container.yml
Comment (RMCrean) [reply .github/workflows/data-builder-container.yml]: Fair point, I've now added this to each workflow file to be more clear in commit 807fa1b.

File: .github/workflows/hugo-container.yml
Comment (RMCrean) [reply .github/workflows/hugo-container.yml:1]: The way I know for this is using a matrix strategy (see e.g. here https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/running-variations-of-jobs-in-a-workflow), it certainly does reduce duplication but I do think it makes it a bit harder to read. Happy to consider changing though in another PR if needed. I know [serve uses this approach to build there images](https://github.com/ScilifelabDataCentre/serve/blob/develop/.github/workflows/publish.yaml)

File: .github/workflows/data-builder-container.yml
Comment (RMCrean) [reply .github/workflows/data-builder-container.yml:3]: Great idea, [this is quite straightforward to setup which is nice](https://docs.github.com/en/actions/writing-workflows/workflow-syntax-for-github-actions#onpushpull_requestpull_request_targetpathspaths-ignore), and I have something similar in the lighthouse.yml gh action already. 

I've now added a setting to prevent rebuilds if only the README.md file is updated in commit 3cc2e149c64c227e84d2b972ef30c3bd7c9a22ee. This could be made better after a good think about exactly what files to commit but I think this should suffice for now.

File: .github/workflows/data-builder-container.yml
Comment (RMCrean) [reply .github/workflows/data-builder-container.yml:68]: It's a bit weird that it seems to be needed but [when I tested on my own testing repo](https://github.com/RMCrean/multi-docker-action/commit/40ef254274076bfb678a5a4a57ac9d96c38574ea) (to prevent polluting this repo with failed builds/commits) it broke the build without it. 

I think this is because the docker/metadata-action@v5 we were using before handled this automatically, whereas with the custom bit of logic in Set Docker tag there is no automatic handling.

**Comment by RMCrean** [comment]
Thanks @kwentine I will merge now even if we have some non-blocking comments/ideas still going so I can finish off work on the rename of the repo. :)

**Pull request merged**
